android/src/main/java/io/ably/flutter/plugin/push/ManualFlutterApplicationRunner.java

```diff
@@ -21,20 +20,27 @@
 import io.flutter.plugin.common.MethodChannel;
 import io.flutter.plugin.common.StandardMethodCodec;
 
-public class PushBackgroundIsolateRunner implements MethodChannel.MethodCallHandler {
-  private static final String TAG = PushBackgroundIsolateRunner.class.getName();
-  private static final String SHARED_PREFERENCES_KEY = "io.ably.flutter.plugin.push.PushBackgroundIsolate.SHARED_PREFERENCES_KEY";
-  private static final String BACKGROUND_MESSAGE_HANDLE_KEY = "BACKGROUND_MESSAGE_HANDLE_KEY";
+/**
+ * This class is used when the application was terminated when the push notification is received.
+ * See [PushMessagingEventHandlers.java] to see where push notifications being handled whilst the
+ * app is in the background or the foreground.
+ *
+ * This class can generified to launch the app manually for any purpose, but currently it
+ * launches the users app and sends it a RemoteMessage.
+ */
+public class ManualFlutterApplicationRunner implements MethodChannel.MethodCallHandler {
+  private static final String TAG = ManualFlutterApplicationRunner.class.getName();
   private final FirebaseMessagingReceiver broadcastReceiver;
   private final RemoteMessage remoteMessage;
   private final MethodChannel backgroundMethodChannel;
 
   @NonNull
   private final FlutterEngine flutterEngine;
-
-  public PushBackgroundIsolateRunner(Context context, FirebaseMessagingReceiver receiver, RemoteMessage message) {
+  // Launch the users Flutter app and pass it the RemoteMessage.
+  // Use this when noo existing Flutter Activity is running.
+  public ManualFlutterApplicationRunner(Context context, FirebaseMessagingReceiver receiver, Intent intent) {
     this.broadcastReceiver = receiver;
-    this.remoteMessage = message;
+    this.remoteMessage = new RemoteMessage(intent.getExtras());
```

Comment:
What if intent or intent.getExtras is null? You might want to annotate non-null fields with @NonNull

Reply:
If the extras is null or is wrongly structured, the remote message **still** gets created, but calling `getNotification` or `getData` leads to a NPE.

This won't happen because the FirebaseMessagingReceiver only receives intents with a specific action. There is no easy way to confirm the intent is structured correctly to add additional safety. But IMHO this case will never happen.